src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,416 @@
+    }
+
+    private void insertIntoKv(Object key, Object value) {
+        kvStore[kvStoreInsertionPos] = key;
+        kvStore[kvStoreInsertionPos+1] = value;
+        kvStoreInsertionPos += 2;
+    }
+
+    private void resizeIndexIfNeeded(int size) {
+        float loadFactor = ((float) (2 * size))/index.length;
+        if (loadFactor < THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD) return;
+
+        int[] oldIndex = this.index;
+        this.index = new int[2*index.length];
+
+        for (int i = 0; i < oldIndex.length; i += 2) {
+            int hash = oldIndex[i];
+            int kvPos = oldIndex[i+1];
+
+            insertIntoIndex(hash,kvPos);
+        }
+    }
+    private void resizeKvStoreIfNeeded() {
+        if (kvStoreInsertionPos < kvStore.length) return;
+
+        kvStore = Arrays.copyOf(kvStore,2*kvStore.length);
+    }
+
+    private CompactHashStore deepCopy(int size) {
+        CompactHashStore copy = new CompactHashStore(size);
+        for (int i = 0; i < index.length; i += 2) {
+            int kvIndex = index[i+1];
+            boolean slotFilled = index[i+1] > INDEX_SLOT_EMPTY_MARKER;
+            if (slotFilled) {
+                kvIndex--; //to account for the +1 offset !
+
+                copy.insertIntoKv(kvStore[kvIndex],kvStore[kvIndex+1]);
+                int hash = index[i];
+                copy.insertIntoIndex(hash,kvIndex+1);
+            }
+        }
+        return copy;
+    }
+
+    @GenerateUncached
+    //Or -1 if key doesn't exist
+    abstract static class GetKeyPosInKVStoreNode extends RubyBaseNode {
+        public abstract int execute(Object key, int hash, int hashPosInIndex, boolean compareByIdentity,
+                                    int[] index, Object[] kvStore);
+
+        @Specialization
+        protected int getKeyPos(Object key, int hash, int hashPosInIndex, boolean compareByIdentity,
+                                int[] index, Object[] kvStore,
+                                @Cached CompareHashKeysNode compareHashKeysNode) {
+            if (hashPosInIndex == -1) return -1;
+
+            int kvPos = index[hashPosInIndex+1] - 1;
+            Object otherKey = kvStore[kvPos];
+
+            //identity implies equality
+            if (otherKey == key) return kvPos;
+            //but non-identity requires we test equality
+            int otherHash = index[hashPosInIndex];
+            if (compareHashKeysNode.execute(compareByIdentity,key,hash,otherKey,otherHash)) {
+                return kvPos;
+            }
+
+            //Key doesn't exist, another one whose hash collides with its hash does
+            return -1;
+        }
+    }
+    @GenerateUncached
+    //Or -1 if hash doesn't exist
+    abstract static class GetHashPosInIndexNode extends RubyBaseNode {
+        public abstract int execute(int hash, int[] index);
+
+        @Specialization
+        protected int getHashPos(int hash, int[] index) {
+            int pos = Math.abs(hash % index.length) & 0xFF_FF_FF_FE;
```

Comment:
I'm thinking if the `index` length is always a power of two we could use `& (index.length-1)` instead of `% index.length`, `%` is as expensive as division.

Also is the `Math.abs` needed or is the mask sufficient to make it positive?

Reply:
Index.length is not always a power of two, it's just always even. I can enforce that it's always a power of 2, but I feel that's pretty memory-inefficient at larger sizes, the closest power of 2 to 1025 is 2048 so we're potentially wasting 1/2 of the index, and it only gets worse from there. 

I searched SO and Hacker's Delight for a bit and there doesn't seem to be an efficient shortcut for even numbers.

Reply:
Yes, make it a power of 2. Anyway hash tables typically double or more their capacity on resize, so it's not worse.